### PR TITLE
CSI-4297 add flashcopy_2 parameter to storgaeclass and volumesnapshotclass

### PR DESCRIPTION
**Pull request by ArbelNathan**

### CODE REVIEW

File: controller/array_action/array_mediator_svc.py [hunk 1]
```diff
@@ -536,14 +536,17 @@ def delete_volume(self, volume_id):
...
+    def get_snapshot(self, volume_id, snapshot_name, pool, flashcopy_2):
         logger.debug("Get snapshot : {}".format(snapshot_name))
-        if self._is_addsnapshot_supported():
-            cli_snapshot = self._get_cli_snapshot_by_name(snapshot_name)
-            if not cli_snapshot:
-                return None
-            source_cli_volume = self._get_cli_volume_by_wwn(volume_id)
-            return self._generate_snapshot_response_from_cli_snapshot(cli_snapshot, source_cli_volume)
+        if flashcopy_2:
+            if self._is_addsnapshot_supported():
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:542]:
```suggestion
        if flashcopy_2 and self._is_addsnapshot_supported():
```

Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py:542]: not good for the raise.. it "if `flashcopy_2`  but not `self._is_addsnapshot_supported()`" - raise

Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:542]: got it

File: controller/array_action/messages.py [hunk 2]
```diff
@@ -14,6 +14,8 @@
 
 SPACE_EFFICIENCY_NOT_SUPPORTED_MESSAGE = "space efficiency is not supported : {0} "
 
+FLASHCOPY_2_NOT_SUPPORTED_MESSAGE = "flashcopy 2.0 is enabled but not supported with object : {0} "
```
Comment (oriyarde) [inline controller/array_action/messages.py:17]: shouldn't it be `with storage : {0}` (and pass the management address instead)?

Comment (ArbelNathan) [reply controller/array_action/messages.py:17]: not necessarily, in #502 we change the check for the source as well, and fcmap in source would raise it too, so it's object-based

File: controller/controller_server/utils.py [hunk 3]
```diff
@@ -102,12 +108,16 @@ def get_object_parameters(parameters, prefix_param_name, system_id):
     default_prefix = parameters.get(prefix_param_name)
     default_io_group = parameters.get(config.PARAMETERS_IO_GROUP)
     default_volume_group = parameters.get(config.PARAMETERS_VOLUME_GROUP)
+    default_flashcopy_2 = parameters.get(config.PARAMETERS_VOLUME_FLASHCOPY_2)
+    flashcopy_2_str = system_parameters.get(config.PARAMETERS_VOLUME_FLASHCOPY_2, default_flashcopy_2)
+    flashcopy_2_bool = _str_to_bool(flashcopy_2_str)
```
Comment (oriyarde) [inline controller/controller_server/utils.py:113]:
```suggestion
    is_flashcopy_2 = yaml.safe_load(raw_flashcopy_2)
```

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 4]
```diff
@@ -441,11 +446,15 @@ def _prepare_mocks_for_get_snapshot_lsvolumesnapshot(self):
...
         self._prepare_mocks_for_get_snapshot_lsvolumesnapshot()
-        snapshot = self.svc.get_snapshot("volume_id", "snapshot_name")
+        snapshot = self.svc.get_snapshot("volume_id", "snapshot_name", None, True)
         self.assertEqual("snapshot_name", snapshot.name)
         self.svc.client.svcinfo.lsvolumesnapshot.assert_called_once_with(filtervalue='snapshot_name=snapshot_name')
         self.svc.client.svcinfo.lsvdisk.assert_called_once_with(bytes=True, filtervalue='vdisk_UID=volume_id')
 
+    def test_get_snapshot_lsvolumesnapshot_not_supported_error(self):
+        with self.assertRaises(array_errors.Flashcopy2NotSupportedMessage):
+            self.svc.get_snapshot("volume_id", "snapshot_name", None, True)
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:456]:
```suggestion
            self.svc.get_snapshot("volume_id", "snapshot_name", pool=None, flashcopy_2=True)
```

File: deploy/kubernetes/examples/demo-storageclass-config-secret.yaml [hunk 5]
```diff
@@ -8,10 +8,11 @@ parameters:
   # non-csi.storage.k8s.io parameters may be specified in by_management_id per system and/or outside by_management_id as the cross-system default.
 
   by_management_id: '{"demo-management-id-1":{"pool":"demo-pool-1","SpaceEfficiency":"dedup_compressed","volume_name_prefix":"demo-prefix-1"},
-                      "demo-management-id-2":{"pool":"demo-pool-2","volume_name_prefix":"demo-prefix-2", "io_group": "demo-iogrp"}}'  # Optional.
+                      "demo-management-id-2":{"pool":"demo-pool-2","volume_name_prefix":"demo-prefix-2", "io_group": "demo-iogrp","flashcopy_2":"false"}}'  # Optional.
   pool: demo-pool
   SpaceEfficiency: thin            # Optional.
   volume_name_prefix: demo-prefix  # Optional.
+  flashcopy_2: "false"               # Optional. Values true/false. The default is false.
```
Comment (oriyarde) [inline deploy/kubernetes/examples/demo-storageclass-config-secret.yaml:15]:
```suggestion
  volume_name_prefix: demo-prefix  # Optional.
  flashcopy_2: "false"             # Optional. Values true/false. The default is false.

```
is it misaligned or is it just GitHub?

File: deploy/kubernetes/examples/demo-volumesnapshotclass-config-secret.yaml [hunk 6]
```diff
@@ -8,10 +8,11 @@ parameters:
   # non-csi.storage.k8s.io parameters may be specified in by_management_id per system and/or outside by_management_id as the cross-system default.
 
   by_management_id: '{"demo-management-id-1":{"pool":"demo-pool-1","SpaceEfficiency":"dedup_compressed","snapshot_name_prefix":"demo-prefix-1"},
-                      "demo-management-id-2":{"pool":"demo-pool-2","snapshot_name_prefix":"demo-prefix-2"}}'  # Optional.
+                      "demo-management-id-2":{"pool":"demo-pool-2","snapshot_name_prefix":"demo-prefix-2","flashcopy_2":"false"}}'  # Optional.
   pool: demo-pool                    # Optional. Use to create the snapshot on a different pool than the source.
   SpaceEfficiency: thin              # Optional. Use to create the snapshot with a different space efficiency than the source.
   snapshot_name_prefix: demo-prefix  # Optional.
+  flashcopy_2: "false"               # Optional. Values true/false. The default is false.
```
Comment (oriyarde) [inline deploy/kubernetes/examples/demo-volumesnapshotclass-config-secret.yaml:15]: similar to io/volume groups, I don't think we need to keep adding every new parameter in this cumbersome topology example

File: controller/array_action/messages.py (hunk 2 above)
Comment (oriyarde) [reply controller/array_action/messages.py:17]: I see. it's hard to keep track when both PRs are open

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 4 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:456]: why?

File: controller/controller_server/utils.py (hunk 3 above)
Comment (ArbelNathan) [reply controller/controller_server/utils.py]: done

File: deploy/kubernetes/examples/demo-storageclass-config-secret.yaml (hunk 5 above)
Comment (ArbelNathan) [reply deploy/kubernetes/examples/demo-storageclass-config-secret.yaml]: misaligned, done.

File: deploy/kubernetes/examples/demo-volumesnapshotclass-config-secret.yaml (hunk 6 above)
Comment (ArbelNathan) [reply deploy/kubernetes/examples/demo-volumesnapshotclass-config-secret.yaml]: ok, done

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 4 above)
Comment (oriyarde) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:456]: the meaning of the hardcoded values is not clear without a name

File: controller/controller_server/utils.py (hunk 3 above)
Comment (oriyarde) [reply controller/controller_server/utils.py]: what about the `yaml.safe_load` part?

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 4 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: deploy/kubernetes/examples/demo-volumesnapshotclass-config-secret.yaml (hunk 6 above)
Comment (oriyarde) [reply deploy/kubernetes/examples/demo-volumesnapshotclass-config-secret.yaml]: same for `demo-storageclass-config-secret.yaml`

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 4 above)
Comment (oriyarde) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: LGTM

File: deploy/kubernetes/examples/demo-volumesnapshotclass-config-secret.yaml (hunk 6 above)
Comment (ArbelNathan) [reply deploy/kubernetes/examples/demo-volumesnapshotclass-config-secret.yaml]: done

**Review by oriyarde — APPROVED** [review]
LGTM

**Review by matancarmeli7 — APPROVED** [review]
LGTM

**Pull request merged**